lib/src/iterable_extensions.dart

```diff
@@ -0,0 +1,725 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/src/utils.dart';
+
+import 'algorithms.dart';
+
+/// Extensions that apply to all iterables.
+///
+/// These extensions provide direct access to some of the
+/// algorithms expose by this package,
+/// as well as some generally useful convenience methods.
+///
+/// More specialized extension methods that only apply to
+/// iterables with specific element types include those of
+/// [IterableComparableExtension] and [IterableNullableExtension].
+extension IterableExtension<T> on Iterable<T> {
+  /// The elements that do not satisfy [test].
+  Iterable<T> whereNot(bool Function(T element) test) =>
+      where((element) => !test(element));
+
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the [compare] [Comparator].
+  List<T> sorted(Comparator<T> compare) => [...this]..sort(compare);
+
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the natural ordering of the
+  /// property [keyOf] of the element.
+  List<T> sortedBy<K extends Comparable<K>>(K Function(T element) keyOf) {
+    var elements = [...this];
+    quickSortBy<T, K>(elements, keyOf, compareComparable);
+    return elements;
+  }
+
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the [compare] [Comparator] of the
+  /// property [keyOf] of the element.
+  List<T> sortedByCompare<K>(
+      K Function(T element) keyOf, Comparator<K> compare) {
+    var elements = [...this];
+    quickSortBy<T, K>(elements, keyOf, compare);
+    return elements;
+  }
+
+  /// Whether the elements are sorted by the [compare] ordering.
+  bool isSorted(Comparator<T> compare) {
+    var iterator = this.iterator;
+    if (!iterator.moveNext()) return true;
```

Comment:
[nit] I think it would be worth calling out the behavior for an empty list in the doc comment.